### PR TITLE
HPCC-19455 Introduce TAKspillread and TAKspillwrite to aid graph viewer

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
Currently pushed for smoke test

**Comment by ghalliday** [comment]
@jakesmith and @shamser please review

File: ecl/hqlcpp/hqlhtcpp.cpp [hunk 1]
```diff
@@ -10683,15 +10685,10 @@ ABoundActivity * HqlCppTranslator::doBuildActivityOutput(BuildCtx & ctx, IHqlExp
...
         s.append(getActivityText(kind));
         s.append("\n");
-        if (expr->hasAttribute(_spill_Atom))
-            s.append("Spill File");
-        else
-        {
-            if (expr->hasAttribute(_workflowPersist_Atom))
-                s.append("Persist ");
+        if (expr->hasAttribute(_workflowPersist_Atom))
+            s.append("Persist ");
```
Comment (shamser) [inline ecl/hqlcpp/hqlhtcpp.cpp]: Should a label be generated for spillwrite ?

File: ecl/hqlcpp/hqlsource.cpp [hunk 2]
```diff
@@ -1802,13 +1803,15 @@ ABoundActivity * SourceBuilder::buildActivity(BuildCtx & ctx, IHqlExpression * e
     translator.gatherActiveCursors(ctx, parentCursors);
 
     bool isSpill = tableExpr && tableExpr->hasAttribute(_spill_Atom);
+    if (isSpill && (activityKind == TAKdiskread))
+        activityKind = TAKspillread;
```
Comment (shamser) [inline ecl/hqlcpp/hqlsource.cpp]: Under what circumstances will there be a spilled read where activityKind != TAKdiskread.  Why is this necessary?

File: ecl/hqlcpp/hqlsource.cpp [hunk 3]
```diff
@@ -1940,7 +1943,10 @@ ABoundActivity * SourceBuilder::buildActivity(BuildCtx & ctx, IHqlExpression * e
     if (nameExpr && nameExpr->queryValue())
     {
         if (isSpill)
-            graphLabel.append("\nSpill");
+        {
+            if (activityKind != TAKspillread)
+                graphLabel.append("\nSpill");
```
Comment (shamser) [inline ecl/hqlcpp/hqlsource.cpp]: Is this here so that diskcount and diskexists on spills are post-fixed to the graph label?

File: roxie/ccd/ccdquery.cpp [hunk 4]
```diff
@@ -593,6 +593,7 @@ class CQueryFactory : implements IQueryFactory, implements IResourceContext, pub
             else
                 return createRoxieServerDiskReadActivityFactory(id, subgraphId, *this, helperFactory, kind, node, remoteId);
         }
+        case TAKspillread:
```
Comment (shamser) [inline roxie/ccd/ccdquery.cpp]: Could possible common up this case with the previous case section.

File: roxie/ccd/ccdserver.cpp [hunk 5]
```diff
@@ -11965,7 +11965,7 @@ class CRoxieServerDiskWriteActivityFactory : public CRoxieServerMultiOutputFacto
         Owned<IHThorDiskWriteArg> helper = (IHThorDiskWriteArg *) helperFactory();
         isTemp = (helper->getFlags() & TDXtemporary) != 0;
         setNumOutputs(helper->getTempUsageCount());
-        if (_kind!=TAKdiskwrite)
+        if (_kind != TAKspillwrite)
```
Comment (shamser) [inline roxie/ccd/ccdserver.cpp]: Should this also check (_kind!=TAKdiskwrite)?

**Review by shamser — COMMENTED** [review]
@ghalliday In SourceBuilder::buildActivity (line 1876-1877) Roxie spill reads will not have have the "Spill" added to the label - is that correct?

File: ecl/hqlcpp/hqlsource.cpp (hunk 3 above)
Comment (ghalliday) [reply ecl/hqlcpp/hqlsource.cpp]: yes

File: ecl/hqlcpp/hqlsource.cpp (hunk 2 above)
Comment (ghalliday) [reply ecl/hqlcpp/hqlsource.cpp]: diskcounts, aggregates and normalize

File: ecl/hqlcpp/hqlhtcpp.cpp (hunk 1 above)
Comment (ghalliday) [reply ecl/hqlcpp/hqlhtcpp.cpp]: no, it already says "Spill Write2  "Spill write Spill File" is too verbose.  It should probably remove the \n as well though - I will fix.

File: ecl/hqlcpp/hqlsource.cpp (hunk 2 above)
Comment (ghalliday) [reply ecl/hqlcpp/hqlsource.cpp]: disk counts, aggregates, normalize

File: ecl/hqlcpp/hqlsource.cpp (hunk 3 above)
Comment (ghalliday) [reply ecl/hqlcpp/hqlsource.cpp]: yes

File: roxie/ccd/ccdquery.cpp (hunk 4 above)
Comment (ghalliday) [reply roxie/ccd/ccdquery.cpp]: this is the correct action to take, so I don't see any benefit.

File: roxie/ccd/ccdserver.cpp (hunk 5 above)
Comment (ghalliday) [reply roxie/ccd/ccdserver.cpp]: no - as far as I know numOutputs should only be non 0 if it is a spill

File: ecl/hqlcpp/hqlhtcpp.cpp (hunk 1 above)
Comment (ghalliday) [reply ecl/hqlcpp/hqlhtcpp.cpp]: Actually it is correct as it is - spilled writes will not go through this branch

**Comment by ghalliday** [comment]
@jakesmith please check

File: common/thorhelper/commonext.cpp [hunk 6]
```diff
@@ -208,6 +208,8 @@ MODULE_INIT(INIT_PRIORITY_STANDARD)
     kindArray[TAKsmartdenormalizegroup] = "smartdenormalizegroup";
     kindArray[TAKselfdenormalize] = "selfdenormalize";
     kindArray[TAKselfdenormalizegroup] = "selfdenormalizegroup";
+    kindArray[TAKspillread] = "diskread" ;
+    kindArray[TAKspillwrite] = "diskwrite" ;
```
Comment (jakesmith) [inline common/thorhelper/commonext.cpp]: Can these be "spilldiskread" and "spilldiskwrite", or do they have to have the same text labels?

Comment (ghalliday) [reply common/thorhelper/commonext.cpp]: I'll change

**Comment by ghalliday** [comment]
Made Jake's suggested change, squashed and repushed.
@richardkchapman please review/merge.
